development/auto-changelog.sh

```diff
@@ -1,26 +1,63 @@
-#! /bin/bash
-# update tags
+#!/usr/bin/env bash
+
+set -e
+set -u
+set -o pipefail
+
+readonly URL='https://github.com/MetaMask/metamask-extension'
+
 git fetch --tags
-# get origin
-URL='https://github.com/MetaMask/metamask-extension'
-# get git logs from last tag until HEAD, pretty by 'subject::body' filtered by grep for PRs made with Github squash merge or Github regular merge
-LOG=$(git log $(git describe --tags $(git rev-list --tags --max-count=1))..HEAD --pretty="%s::%b" --reverse --grep="Merge pull request #" --grep="(#");
-while read -r line; do
-    # get git log subject
-    SUBJECT=$(echo $line | sed -E 's/(.*):{2}(.*)/\1/')
-    # get git log PR id, PR made with Github squash merge or Github regular merge
-    PR=$(echo $SUBJECT | sed 's/^.*(#\([^&]*\)).*/\1/' | sed 's/^.*#\([^&]*\) from.*/\1/')
-    # if PR made with Github squash merge, subject is the body
-    if [ -z "$(echo $line | sed -E 's/(.*):{2}(.*)/\2/')" ]; then
-        BODY=$(echo $SUBJECT | sed "s/(#$PR)//g"); else
-        BODY=$(echo $line | sed -E 's/(.*):{2}(.*)/\2/') 
+
+most_recent_tag="$(git describe --tags "$(git rev-list --tags --max-count=1)")"
+
+readarray -t commits <<< "$(git rev-list "${most_recent_tag}"..HEAD)"
+
+for commit in "${commits[@]}"
+do
+    subject="$(git show -s --format="%s" "$commit")"
+
+    # Parse a commit for the PR ID and a description. Three types of commits
+    # are handled: Squash & Merge, Merge, and normal commits.
+    # Squash & Merge: the commit subject is parsed as `<description> (#<PR ID>)`
+    # Merge: the PR ID is parsed from the git subject (which is of the form `Merge pull request
+    #   #<PR ID> from <branch>`, and the description is assumed to be the first line of the body.
+    #   If no body is found, the description is set to the commit subject
+    # Normal commits: The commit subject is the description, and the PR ID is omitted.
+    if grep -E -q '\(#[[:digit:]]+\)|#[[:digit:]]+\sfrom' <<< "$subject"
+    then
+        # shellcheck disable=SC2001
+        pr="$(printf "%s" "$subject" | sed 's/^.*#\([[:digit:]]\+\).*$/\1/')"
```

Comment:
This doesn't quite work on macOS. 

`GNU bash, version 4.4.19(1)-release (x86_64-pc-linux-gnu)`:

```
# subject='Merge pull request #6745 from MetaMask/fix-readme'
# pr="$(printf "%s" "$subject" | sed 's/^.*#\([[:digit:]]\+\).*$/\1/')"
# printf '%s\n' "$pr"
6745
```

`GNU bash, version 5.0.0(1)-release (x86_64-apple-darwin17.7.0)`:

```
$ subject='Merge pull request #6745 from MetaMask/fix-readme'
$ pr="$(printf "%s" "$subject" | sed 's/^.*#\([[:digit:]]\+\).*$/\1/')"
$ printf '%s\n' "$pr"
Merge pull request #6745 from MetaMask/fix-readme
```

We might be able to use AWK here instead of sed? The following works on macOS:

```suggestion
        pr="$(awk '{sub("#", "", $4); print $4}' <<< $subject)"
```

Reply:
Although below we're using `#$pr` twice and `$pr` once so we could maybe switch its value to having the `#` prefix (just `awk '{print $4}' <<< $subject`) and then stripping it (`"${pr###}"`) when we don't need it.

Reply:
This line is meant to grab the PR id from both "Squash & Merge" and "Merge" cases though. For "Squash & Merge" the _last_ field has the id rather than the fourth.

Reply:
I think it would be cleaner to deal with those two cases separately. Then we could use `awk` as you suggested, which seems far simpler than whatever's going on with `sed`.

I was surprised to see `sed` fail like that - maybe it's the escape character? The default on Linux is to escape special characters, but there's a flag to invert that. Maybe the inverse is default on macOS.